### ISSUE TITLE
API: embedded object's property of reference type return actual DB id in the API response

### ISSUE THREAD

**Issue by sangeethah** [rancher/rancher#594]
So "service" object has embedded launchConfig property of "container" type. If the embedded object has properties that are ref to the other cattle object - dataVolumesFrom/registryCredentialId in my case - their ids in the response object are listed as actual DB ids in the API response: 22 instead of 1c22 as an example

**Comment by sangeethah** [comment]
When we activate services with images from private registry - quay.io which requires credentials, registryCredentialId stored in rancher.yml:

testp1:
scale: 1
registryCredentialId: 15

In this case , we do not expect the DB ids to be exposed.

**Comment by alena1108** [comment]
should be fixed as a part of v2 API

**Comment by prachidamle** [comment]
@sangeethah I tried reproducing the issue for registryCredentialId , but I see that the Id is obfuscated and not a D ID anymore. Please reopen if you happen to see this again.

{
"id": "1i125",
"type": "container",
"links": { … },
"actions": { … },
"name": "testregistry_Q4_1",
"state": "running",
"accountId": "1a5",
"build": null,
"capAdd": [ ],
"capDrop": [ ],
"command": null,
"cpuSet": null,
"cpuShares": null,
"createIndex": 1,
"created": "2016-03-09T18:20:25Z",
"createdTS": 1457547625000,
"dataVolumeMounts": { },
"dataVolumes": [ ],
"dataVolumesFrom": [ ],
"dataVolumesFromLaunchConfigs": [ ],
"deploymentUnitUuid": "2b5b75fb-469f-489f-9c20-8acc69912a73",
"description": null,
"devices": [ ],
"dns": [ … ],
"dnsSearch": [ 2 items … ],
"domainName": null,
"entryPoint": null,
"environment": null,
"expose": null,
"externalId": "2c7488629ac5991782b02cdc18ca554168a3bc0ab870411e8dd42abc7cbe1635",
"extraHosts": null,
"firstRunning": "2016-03-09T18:20:56Z",
"firstRunningTS": 1457547656000,
"healthCheck": null,
"healthState": null,
"hostname": null,
"imageUuid": "docker:quay.io/sangeetha/test1",
"kind": "container",
"labels": { … },
"logConfig": null,
"lxcConf": null,
"memory": null,
"memorySwap": null,
"nativeContainer": false,
"networkContainerId": null,
"networkMode": "managed",
"pidMode": null,
"ports": [ ],
"primaryIpAddress": "10.42.109.135",
"privileged": false,
"publishAllPorts": false,
"readOnly": false,
"registryCredentialId": "1c68",
....
}

**Comment by prachidamle** [comment]
This is how to reproduce this:
1) Run integration test: test_activate_single_service from test_svc_discovery.py
2) Change the account_id to 5 for the services and environment the test creates in the DB
3) In API, list the service and check the dataVolumesFrom
4) It exposes the container's Db Id

"dataVolumesFrom": [
5,
],

**Comment by prachidamle** [comment]
@alena1108 mentioned this is not critical and will be fixed along with v2 API. Hence punting it to future.

**Comment by alena1108** [comment]
Looks like we have to fix it for GA as it affects rancher-compose/UI

https://github.com/rancher/rancher/issues/3970

**Comment by wayneconnolly** [comment]
Great, so what is the eta?

**Comment by deniseschannon** [comment]
GA is expected for end of March. :)

**Comment by wayneconnolly** [comment]
Great. Thank-you. Will monitor this issue.

**Comment by sangeethah** [comment]
Tested with latest build from master:

Now we do not see ids being used directly in dataVolumesFrom/requestedHostId/registryCredentialId

```
"dataVolumesFrom": [
"1i1659",
],
```

```
"registryCredentialId": "1c155",
```

```
"requestedHostId": "1h4",
```

**Comment by alena1108** [comment]
The bug still happens when an embedded type is list. So if Object 1 has a field []Object2 (list), and Object2 references Object3, Object's 3 id is not obfuscated.

**Comment by ibuildthecloud** [comment]
@alena1108 can you give me an example of this?  which API?  I'll fix it as I'm working in this area already.

**Comment by alena1108** [comment]
@ibuildthecloud it used to be actually for the serviceId field under balancerService.portRules[] (before I removed the type). You can get it by rebasing to this commit of the lb refactor PR:

https://github.com/rancher/cattle/pull/1798/commits/ef9bcd03d98e9e64549a43ef3e8cc88f9f617b65

**Comment by VAdamec** [comment]
@deniseschannon is there any progress with that ? You closed #3951 as duplicate of #3929, than #3929 in favor of this :-) but it's still issue
